app/src/main/java/org/schabi/newpipe/player/playback/TrackSelectionListener.java

```diff
@@ -0,0 +1,106 @@
+     *                        null
+     * @param rendererIndex   the index of the video renderer
+     * @param qualityTextView the {@link TextView quality text view} on which the playing quality
+     *                        will be updated after a quality item is clicked
+     * @throws NullPointerException if trackSelector is null
+     */
+    public TrackSelectionListener(@NonNull final Context context,
+                                  @NonNull final CustomTrackSelector trackSelector,
+                                  final int rendererIndex,
+                                  @NonNull final TextView qualityTextView) {
+        this.context = context;
+        this.trackSelector = trackSelector;
+        final MappingTrackSelector.MappedTrackInfo mappedTrackInfo = checkNotNull(
+                trackSelector.getCurrentMappedTrackInfo());
+        this.rendererIndex = rendererIndex;
+        this.qualityTextView = qualityTextView;
+
+        rendererTrackGroups = mappedTrackInfo.getTrackGroups(rendererIndex);
+    }
+
+    @Override
+    public boolean onMenuItemClick(@NonNull final MenuItem item) {
+        final DefaultTrackSelector.Parameters selectionParameters = trackSelector.getParameters();
+        final boolean isDisabled = selectionParameters.getRendererDisabled(rendererIndex);
+        final DefaultTrackSelector.ParametersBuilder builder = selectionParameters.buildUpon()
+                .clearSelectionOverrides(rendererIndex)
+                .setRendererDisabled(rendererIndex, isDisabled);
+        // The auto string
+        final String qualityTextViewText = qualityTextView.getText().toString();
+
+        if (item.getItemId() != 0) {
+            // A resolution string
+            final TrackGroup videoTrackGroup = rendererTrackGroups.get(0);
+
+            final int qualityId = videoTrackGroup.length - item.getItemId();
+            final DefaultTrackSelector.SelectionOverride selectionOverride =
+                    new DefaultTrackSelector.SelectionOverride(
+                            rendererIndex, qualityId);
+            builder.setSelectionOverride(rendererIndex, rendererTrackGroups,
+                    selectionOverride);
+
+            // The quality needs only to be set here when a user keeps the quality selected
+            // automatically by ExoPlayer but selects it manually with the quality selector (the
+            // quality string will be changed in other cases with the analytics listener set in the
+            // player)
+            if (qualityTextViewText.contains(context.getString(R.string.auto_quality))) {
+                final Format currentPlayingFormat = videoTrackGroup.getFormat(qualityId);
+                qualityTextView.setText(getResolutionStringFromFormat(context,
+                        currentPlayingFormat));
+            }
+
+            trackSelector.setParameters(builder);
+        } else {
+            trackSelector.setParameters(builder);
```

Comment:
Looks similar, maybe deduplicateable?

Reply:
I think we can't, otherwise the previous quality played will be set as the quality currently played (because the event listener which update the quality will be not called) so I think this will lead to an issue in a specific case.